src/Controller/SubmitController.php

```diff
@@ -32,6 +33,36 @@ public function redirectAction(Request $request) : Response
 
         $jwt = $this->get('elife.journal.security.xpub.token_generator')->generate($user, $request->getSession()->remove('journal.submit') ?? false);
 
-        return new RedirectResponse("{$this->getParameter('submit_url')}#{$jwt}");
+        // if a return url is specified, check that its from a trusted host
+        $returnUrl = $request->query->get('return_url', null);
+
+        if (is_null($returnUrl)) {
+            $returnUrl = $this->getParameter('submit_url');
+        } else {
+            $trustedHosts = $this->getParameter('trusted_hosts');
+            $trusted = false;
+
+            $uri = new Uri($returnUrl);
+            foreach ($trustedHosts as $trustedHost) {
+                if (preg_match("/{$trustedHost}/", $uri->getHost())) {
+                    $trusted = true;
+                }
+            }
+
+            if (!$trusted) {
+                throw new NotFoundHttpException('Not allowed to see xPub');
+            }
+        }
+
+        $redirectUrl = "{$returnUrl}#{$jwt}";
+
+        // return in query arg if specified
+        $tokenInQueryArg = $request->query->get('token_in_query', false);
```

Comment:
Is this an xPub/Reviewer difference? Weird to have knowledge of it here.

Reply:
Getting back the token inside a query arg means its easier for reviewer to do the token exchange. Currently the reviewer-client needs to do an extra redirection to send the journal token to `continuum-adaptor` and sign a new token for reviewer.

Reply:
Not a fan of having this either. Feels like xPub should be changed to handle either, then switch over.

Reply:
Could also start making the redirect URI required at the same time.

Reply:
Means not having to have both `submit_url` and `submit_url_redirects` here.

Reply:
@thewilkybarkid once xpub is retired and libero reviewer is launched, we can switch to the new behaviour only. I don't think there is any point in changing xpub.